### ISSUE TITLE
[Feature request] Expanding matchmaker criteria on server

### ISSUE THREAD

**Issue by dmitryrn** [heroiclabs/nakama#1064]
It would be useful to be able to expand criteria on server, for example:
```
extracted := nk.MatchmakerExtract()
// see which tickets were in matchmaker for a long time, modify their query and properties;
nk.MatchmakerInsert(extracted)
```

I haven’t tested that yet, but it seems that it is possible to expose this matchmaker functionality to the user code on server. LocalMatchmaker isn’t tied to websocket in a way that prevents using the matchmaker in the user code on the server. As far as I was able to tell.

We’re actually considering forking Nakama because we want to expand matchmaking criteria on the server and not on the client, because the criteria depends on the backend data, so it makes more sense to have the matchmaker expansion logic on the server.

One thing I can think of is clusterisation - multiple instances of Nakama. Here we may have a problem, but I'm not sure how Nakama's clusterisation works.

**Comment by mofirouz** [comment]
I think it would be good to understand the exact use case here:

- can you tell us what information is available on the server that you'd like to use to enhance matchmaking criteria? 
- Also how can this be different to using a BeforeHook on Matchmaker request to add data for that user into the matchmaking criteria before it is processed by Nakama?
- What kind of gameplay are you developing here?

**Comment by mofirouz** [comment]
Also since you've edited the original message,
> One thing I can think of is clusterisation - multiple instances of Nakama. Here we may have a problem, but I'm not sure how Nakama's clusterisation works.

Are you currently using Nakama Enterprise?

**Comment by dmitryrn** [comment]
> Also since you've edited the original message,
> 
> > One thing I can think of is clusterisation - multiple instances of Nakama. Here we may have a problem, but I'm not sure how Nakama's clusterisation works.
> 
> Are you currently using Nakama Enterprise?

No, we're not using Nakama enterprise

**Comment by dmitryrn** [comment]
> I think it would be good to understand the exact use case here:
> 
>     * can you tell us what information is available on the server that you'd like to use to enhance matchmaking criteria?
> 
>     * Also how can this be different to using a BeforeHook on Matchmaker request to add data for that user into the matchmaking criteria before it is processed by Nakama?
> 
>     * What kind of gameplay are you developing here?

1) number of matches played, number of matches won, some kind of player's rank

2)
How it is with the current situation: a player clicks Play and adds their matchmaker entry to the matchmaker, they sit there for X seconds and we know that we should expand the criteria. If we do it on the client, then the logic with timers and requesting matchmaker related data is on the client. We add another ticket with an expanded criteria, which is fine. But we're ideally thinking of adding themselves to the matchmaker once and then getting into the game while all the matchmaking expansion logic is done on the backend.

We also want to validate query and properties that the client has submitted on the backed, via hooks. We can do that but that means that Nakama on the server should also fetch matchmaker related data each time a user adds themselves to matchmaker.

3) a shooter similar to Overwatch

**Comment by mofirouz** [comment]
> But we're ideally thinking of adding themselves to the matchmaker once and then getting into the game while all the matchmaking expansion logic is done on the backend.

Why can't this be done transparently on the client side whilst the player is in the game/match? The action of sending a matchmaker API call doesn't have to be explicitly driven by a player action.

>How it is with the current situation: a player clicks Play and adds their matchmaker entry to the matchmaker, they sit there for X seconds and we know that we should expand the criteria. If we do it on the client, then the logic with timers and requesting matchmaker related data is on the client. We add another ticket with an expanded criteria, which is fine. 

So if I read this, you'd like to mutate the matchmaker criteria _after_ it's been submitted to the matchmaker, without cancelling/resubmitting? Are you sure you'd want to open a PR about this, this is non-trivial to get right.

**Comment by dmitryrn** [comment]
> Why can't this be done transparently on the client side whilst the player is in the game/match?
One problem with this is every time we want to update the matchmaker logic, we'll need to update the client, and push it to users. To mitigate that we'll need to do something like fetching matchmaking configuration from the server.

**Comment by mofirouz** [comment]
>To mitigate that we'll need to do something like fetching matchmaking configuration from the server and passing it to .MatchmakerAdd(query, properties)

I think this is where a remote configuration system, like [Satori](https://heroiclabs.com/satori) comes in to play to be able to have these customisation available per-player without a game client update.

As I said originally, we can take a look at the PR, but this is non-trivial work to add and get right in a highly concurrent environment without significant overhead in performance. I suggest you take a different approach as we've spoken on this issue.

**Comment by dmitryrn** [comment]
Thank you. Another related question: would that be a performance hit or some other problem, if we added localmatchmaker.Extract() method to the user code on the server? Right now matchmaker is like a black box for the server and we can't monitor its state. What, for example, if we want to see if no one is playing a certain game mode and we want to hide that in the client?

As far as I understand, extract just extracts the whole matchmaker state. Would that be fine if we extracted it once every few minutes to make a decision?
